File: android/app/src/main/java/com/xiangshang/youth/core/service/CourseApi.kt
package com.xiangshang.youth.core.service

import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.PUT
import retrofit2.http.Path

data class CaptionTrack(val url: String? = null, val uri: String? = null, val language: String? = null, val label: String? = null, val mimeType: String? = null)
data class RemoteLesson(val courseId: String, val moduleId: String?, val lessonId: String, val title: String, val lessonTitle: String, val durationMs: Int, val videoSource: String?, val lastPositionMs: Int, val completed: Boolean, val version: Int, val captions: List<CaptionTrack> = emptyList())
data class PlaybackSource(val lessonId: String, val courseId: String, val videoSource: String, val durationMs: Int, val captions: List<CaptionTrack> = emptyList(), val playbackSessionId: String? = null, val expiresAt: String? = null, val refreshAfterSeconds: Int? = null)
data class LessonProgressWrite(val lastPositionMs: Int, val completed: Boolean, val expectedVersion: Int? = null)
data class LessonProgressAck(val lessonId: String, val lastPositionMs: Int, val completed: Boolean, val version: Int)
interface CourseApi {
    @GET("v1/students/{studentId}/courses") suspend fun courses(@Path("studentId") studentId: String): ApiEnvelope<List<RemoteLesson>>
    @GET("v1/lessons/{lessonId}/playback") suspend fun playback(@Path("lessonId") lessonId: String): ApiEnvelope<PlaybackSource>
    @PUT("v1/students/{studentId}/lessons/{lessonId}/progress") suspend fun saveProgress(@Path("studentId") studentId: String, @Path("lessonId") lessonId: String, @Body body: LessonProgressWrite): ApiEnvelope<LessonProgressAck>
}

File: ios/XiangshangYouth/Core/Services/CourseApi.swift
import Foundation

struct CourseApi {
    let client: ApiClient
    init(client: ApiClient = .shared) { self.client = client }
    func courses(childID: String) async throws -> [RemoteLesson] { try await client.request(path: "v1/students/\(childID)/courses", type: [RemoteLesson].self) }
    func playback(lessonID: String) async throws -> PlaybackSource { try await client.request(path: "v1/lessons/\(lessonID)/playback", type: PlaybackSource.self) }
    func saveProgress(childID: String, lessonID: String, lastPositionMs: Int, completed: Bool, expectedVersion: Int?) async throws -> LessonProgressAck {
        try await client.request(path: "v1/students/\(childID)/lessons/\(lessonID)/progress", method: "PUT", body: LessonProgressWrite(lastPositionMs: lastPositionMs, completed: completed, expectedVersion: expectedVersion), type: LessonProgressAck.self)
    }
}
struct RemoteLesson: Decodable { let courseID: String; let moduleID: String?; let lessonID: String; let title: String; let lessonTitle: String; let durationMs: Int; let videoSource: String?; let lastPositionMs: Int; let completed: Bool; let version: Int; enum CodingKeys: String, CodingKey { case title, lessonTitle, durationMs, videoSource, lastPositionMs, completed, version; case courseID = "courseId"; case moduleID = "moduleId"; case lessonID = "lessonId" } }
extension RemoteLesson: Identifiable { var id: String { lessonID } }
struct CaptionTrack: Decodable { let url: String?; let uri: String?; let language: String?; let label: String?; let mimeType: String? }
struct PlaybackSource: Decodable { let lessonID: String; let courseID: String; let videoSource: String; let durationMs: Int; let captions: [CaptionTrack]?; let playbackSessionID: String?; let expiresAt: String?; let refreshAfterSeconds: Int?; enum CodingKeys: String, CodingKey { case videoSource, durationMs, captions, expiresAt, refreshAfterSeconds; case lessonID = "lessonId"; case courseID = "courseId"; case playbackSessionID = "playbackSessionId" } }
private struct LessonProgressWrite: Encodable { let lastPositionMs: Int; let completed: Bool; let expectedVersion: Int? }
struct LessonProgressAck: Decodable { let lessonID: String; let lastPositionMs: Int; let completed: Bool; let version: Int; enum CodingKeys: String, CodingKey { case lastPositionMs, completed, version; case lessonID = "lessonId" } }
